File: backend/vessel/domain/models/execution.py
"""Execution layer models - 'What HAS happened' (work items, logs)."""

import uuid
from datetime import datetime
from typing import Any

from sqlalchemy import BigInteger, Boolean, ForeignKey, Integer, String, Text, func
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column, relationship

from vessel.domain.models.base import Base, TimestampMixin


class WorkItem(TimestampMixin, Base):
    __tablename__ = "work_items"

    id: Mapped[uuid.UUID] = mapped_column(
        primary_key=True, default=uuid.uuid4, server_default=func.gen_random_uuid()
    )
    pipeline_activation_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("pipeline_activations.id", ondelete="RESTRICT"), nullable=False
    )
    pipeline_instance_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("pipeline_instances.id", ondelete="RESTRICT"), nullable=False
    )
    source_type: Mapped[str] = mapped_column(String(20), nullable=False)
    source_key: Mapped[str] = mapped_column(String(1024), nullable=False)
    source_metadata: Mapped[dict[str, Any]] = mapped_column(JSONB, nullable=False, server_default="{}")
    dedup_key: Mapped[str | None] = mapped_column(String(512))
    detected_at: Mapped[datetime] = mapped_column(server_default=func.now(), nullable=False)
    status: Mapped[str] = mapped_column(String(20), nullable=False, server_default="DETECTED")
    current_execution_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("work_item_executions.id", ondelete="SET NULL", use_alter=True)
    )
    execution_count: Mapped[int] = mapped_column(Integer, nullable=False, server_default="0")
    last_completed_at: Mapped[datetime | None] = mapped_column()

    pipeline_activation: Mapped["PipelineActivation"] = relationship(  # type: ignore[name-defined]
        back_populates="work_items",
    )
    pipeline_instance: Mapped["PipelineInstance"] = relationship()  # type: ignore[name-defined]
    executions: Mapped[list["WorkItemExecution"]] = relationship(
        back_populates="work_item",
        cascade="all, delete-orphan",
        foreign_keys="WorkItemExecution.work_item_id",
    )
    reprocess_requests: Mapped[list["ReprocessRequest"]] = relationship(
        back_populates="work_item",
        cascade="all, delete-orphan",
    )


class WorkItemExecution(Base):
    __tablename__ = "work_item_executions"

    id: Mapped[uuid.UUID] = mapped_column(
        primary_key=True, default=uuid.uuid4, server_default=func.gen_random_uuid()
    )
    work_item_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("work_items.id", ondelete="CASCADE"), nullable=False
    )
    execution_no: Mapped[int] = mapped_column(Integer, nullable=False)
    trigger_type: Mapped[str] = mapped_column(String(20), nullable=False, server_default="INITIAL")
    trigger_source: Mapped[str | None] = mapped_column(String(256))
    status: Mapped[str] = mapped_column(String(20), nullable=False, server_default="RUNNING")
    started_at: Mapped[datetime] = mapped_column(server_default=func.now(), nullable=False)
    ended_at: Mapped[datetime | None] = mapped_column()
    duration_ms: Mapped[int | None] = mapped_column(BigInteger)
    reprocess_request_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("reprocess_requests.id", ondelete="SET NULL", use_alter=True)
    )
    created_at: Mapped[datetime] = mapped_column(server_default=func.now(), nullable=False)

    work_item: Mapped["WorkItem"] = relationship(
        back_populates="executions",
        foreign_keys=[work_item_id],
    )
    step_executions: Mapped[list["WorkItemStepExecution"]] = relationship(
        back_populates="execution",
        cascade="all, delete-orphan",
    )
    snapshot: Mapped["ExecutionSnapshot | None"] = relationship(
        back_populates="execution",
        uselist=False,
    )
    event_logs: Mapped[list["ExecutionEventLog"]] = relationship(
        back_populates="execution",
        cascade="all, delete-orphan",
    )


class WorkItemStepExecution(Base):
    __tablename__ = "work_item_step_executions"

    id: Mapped[uuid.UUID] = mapped_column(
        primary_key=True, default=uuid.uuid4, server_default=func.gen_random_uuid()
    )
    execution_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("work_item_executions.id", ondelete="CASCADE"), nullable=False
    )
    pipeline_step_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("pipeline_steps.id", ondelete="RESTRICT"), nullable=False
    )
    step_type: Mapped[str] = mapped_column(String(20), nullable=False)
    step_order: Mapped[int] = mapped_column(Integer, nullable=False)
    status: Mapped[str] = mapped_column(String(20), nullable=False, server_default="PENDING")
    started_at: Mapped[datetime | None] = mapped_column()
    ended_at: Mapped[datetime | None] = mapped_column()
    duration_ms: Mapped[int | None] = mapped_column(BigInteger)
    input_summary: Mapped[dict[str, Any] | None] = mapped_column(JSONB)
    output_summary: Mapped[dict[str, Any] | None] = mapped_column(JSONB)
    error_code: Mapped[str | None] = mapped_column(String(128))
    error_message: Mapped[str | None] = mapped_column(Text)
    retry_attempt: Mapped[int] = mapped_column(Integer, nullable=False, server_default="0")
    created_at: Mapped[datetime] = mapped_column(server_default=func.now(), nullable=False)

    execution: Mapped["WorkItemExecution"] = relationship(back_populates="step_executions")
    pipeline_step: Mapped["PipelineStep"] = relationship()  # type: ignore[name-defined]
    event_logs: Mapped[list["ExecutionEventLog"]] = relationship(
        back_populates="step_execution",
    )


class ExecutionSnapshot(Base):
    __tablename__ = "execution_snapshots"

    id: Mapped[uuid.UUID] = mapped_column(
        primary_key=True, default=uuid.uuid4, server_default=func.gen_random_uuid()
    )
    execution_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("work_item_executions.id", ondelete="CASCADE"), unique=True, nullable=False
    )
    pipeline_config: Mapped[dict[str, Any]] = mapped_column(JSONB, nullable=False, server_default="{}")
    collector_config: Mapped[dict[str, Any]] = mapped_column(JSONB, nullable=False, server_default="{}")
    algorithm_config: Mapped[dict[str, Any]] = mapped_column(JSONB, nullable=False, server_default="{}")
    transfer_config: Mapped[dict[str, Any]] = mapped_column(JSONB, nullable=False, server_default="{}")
    snapshot_hash: Mapped[str | None] = mapped_column(String(128))
    created_at: Mapped[datetime] = mapped_column(server_default=func.now(), nullable=False)

    execution: Mapped["WorkItemExecution"] = relationship(back_populates="snapshot")


class ReprocessRequest(TimestampMixin, Base):
    __tablename__ = "reprocess_requests"

    id: Mapped[uuid.UUID] = mapped_column(
        primary_key=True, default=uuid.uuid4, server_default=func.gen_random_uuid()
    )
    work_item_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("work_items.id", ondelete="CASCADE"), nullable=False
    )
    requested_by: Mapped[str] = mapped_column(String(256), nullable=False)
    requested_at: Mapped[datetime] = mapped_column(server_default=func.now(), nullable=False)
    reason: Mapped[str | None] = mapped_column(Text)
    start_from_step: Mapped[int | None] = mapped_column(Integer)
    use_latest_recipe: Mapped[bool] = mapped_column(Boolean, nullable=False, server_default="true")
    status: Mapped[str] = mapped_column(String(20), nullable=False, server_default="PENDING")
    approved_by: Mapped[str | None] = mapped_column(String(256))
    execution_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("work_item_executions.id", ondelete="SET NULL", use_alter=True)
    )

    work_item: Mapped["WorkItem"] = relationship(back_populates="reprocess_requests")


class ExecutionEventLog(Base):
    __tablename__ = "execution_event_logs"

    id: Mapped[uuid.UUID] = mapped_column(
        primary_key=True, default=uuid.uuid4, server_default=func.gen_random_uuid()
    )
    execution_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("work_item_executions.id", ondelete="CASCADE"), nullable=False
    )
    step_execution_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("work_item_step_executions.id", ondelete="CASCADE")
    )
    event_type: Mapped[str] = mapped_column(String(10), nullable=False, server_default="INFO")
    event_code: Mapped[str] = mapped_column(String(128), nullable=False)
    message: Mapped[str | None] = mapped_column(Text)
    detail_json: Mapped[dict[str, Any] | None] = mapped_column(JSONB)
    created_at: Mapped[datetime] = mapped_column(server_default=func.now(), nullable=False)

    execution: Mapped["WorkItemExecution"] = relationship(back_populates="event_logs")
    step_execution: Mapped["WorkItemStepExecution | None"] = relationship(back_populates="event_logs")

File: engine/src/Hermes.Engine/Domain/Entities/Execution.cs
namespace Hermes.Engine.Domain.Entities;

public class WorkItem : AuditableEntity
{
    public Guid PipelineActivationId { get; set; }
    public Guid PipelineInstanceId { get; set; }
    public SourceType SourceType { get; set; }
    public string SourceKey { get; set; } = string.Empty;
    public string SourceMetadata { get; set; } = "{}";
    public string? DedupKey { get; set; }
    public DateTimeOffset DetectedAt { get; set; } = DateTimeOffset.UtcNow;
    public JobStatus Status { get; set; } = JobStatus.Detected;
    public Guid? CurrentExecutionId { get; set; }
    public int ExecutionCount { get; set; }
    public DateTimeOffset? LastCompletedAt { get; set; }
    public PipelineActivation PipelineActivation { get; set; } = null!;
    public PipelineInstance PipelineInstance { get; set; } = null!;
    public List<WorkItemExecution> Executions { get; set; } = new();
    public List<ReprocessRequest> ReprocessRequests { get; set; } = new();
}

public class WorkItemExecution : BaseEntity
{
    public Guid WorkItemId { get; set; }
    public int ExecutionNo { get; set; }
    public TriggerType TriggerType { get; set; } = TriggerType.Initial;
    public string? TriggerSource { get; set; }
    public ExecutionStatus Status { get; set; } = ExecutionStatus.Running;
    public DateTimeOffset StartedAt { get; set; } = DateTimeOffset.UtcNow;
    public DateTimeOffset? EndedAt { get; set; }
    public long? DurationMs { get; set; }
    public Guid? ReprocessRequestId { get; set; }
    public DateTimeOffset CreatedAt { get; set; } = DateTimeOffset.UtcNow;
    public WorkItem WorkItem { get; set; } = null!;
    public List<WorkItemStepExecution> StepExecutions { get; set; } = new();
    public ExecutionSnapshot? Snapshot { get; set; }
    public List<ExecutionEventLog> EventLogs { get; set; } = new();
}

public class WorkItemStepExecution : BaseEntity
{
    public Guid ExecutionId { get; set; }
    public Guid PipelineStepId { get; set; }
    public StageType StepType { get; set; }
    public int StepOrder { get; set; }
    public StepExecutionStatus Status { get; set; } = StepExecutionStatus.Pending;
    public DateTimeOffset? StartedAt { get; set; }
    public DateTimeOffset? EndedAt { get; set; }
    public long? DurationMs { get; set; }
    public string? InputSummary { get; set; }
    public string? OutputSummary { get; set; }
    public string? ErrorCode { get; set; }
    public string? ErrorMessage { get; set; }
    public int RetryAttempt { get; set; }
    public DateTimeOffset CreatedAt { get; set; } = DateTimeOffset.UtcNow;
    public WorkItemExecution Execution { get; set; } = null!;
    public PipelineStep PipelineStep { get; set; } = null!;
    public List<ExecutionEventLog> EventLogs { get; set; } = new();
}

public class ExecutionSnapshot : BaseEntity
{
    public Guid ExecutionId { get; set; }
    public string PipelineConfig { get; set; } = "{}";
    public string CollectorConfig { get; set; } = "{}";
    public string ProcessConfig { get; set; } = "{}";
    public string ExportConfig { get; set; } = "{}";
    public string? SnapshotHash { get; set; }
    public DateTimeOffset CreatedAt { get; set; } = DateTimeOffset.UtcNow;
    public WorkItemExecution Execution { get; set; } = null!;
}

public class ExecutionEventLog : BaseEntity
{
    public Guid ExecutionId { get; set; }
    public Guid? StepExecutionId { get; set; }
    public EventLevel EventType { get; set; } = EventLevel.Info;
    public string EventCode { get; set; } = string.Empty;
    public string? Message { get; set; }
    public string? DetailJson { get; set; }
    public DateTimeOffset CreatedAt { get; set; } = DateTimeOffset.UtcNow;
    public WorkItemExecution Execution { get; set; } = null!;
    public WorkItemStepExecution? StepExecution { get; set; }
}

public class ReprocessRequest : AuditableEntity
{
    public Guid WorkItemId { get; set; }
    public string RequestedBy { get; set; } = string.Empty;
    public DateTimeOffset RequestedAt { get; set; } = DateTimeOffset.UtcNow;
    public string? Reason { get; set; }
    public int? StartFromStep { get; set; }
    public bool UseLatestRecipe { get; set; } = true;
    public ReprocessStatus Status { get; set; } = ReprocessStatus.Pending;
    public string? ApprovedBy { get; set; }
    public Guid? ExecutionId { get; set; }
    public WorkItem WorkItem { get; set; } = null!;
}

public class DeadLetterEntry : BaseEntity
{
    public Guid WorkItemId { get; set; }
    public Guid? ExecutionId { get; set; }
    public Guid PipelineInstanceId { get; set; }
    public string ErrorCode { get; set; } = string.Empty;
    public string ErrorMessage { get; set; } = string.Empty;
    public string? StackTrace { get; set; }
    public int FailureCount { get; set; }
    public string? LastStepType { get; set; }
    public int? LastStepOrder { get; set; }
    public string OriginalSourceKey { get; set; } = string.Empty;
    public string? InputDataJson { get; set; }
    public DeadLetterStatus Status { get; set; } = DeadLetterStatus.Quarantined;
    public string? ResolvedBy { get; set; }
    public DateTimeOffset? ResolvedAt { get; set; }
    public string? ResolutionNote { get; set; }
    public DateTimeOffset CreatedAt { get; set; } = DateTimeOffset.UtcNow;
    public WorkItem WorkItem { get; set; } = null!;
}
